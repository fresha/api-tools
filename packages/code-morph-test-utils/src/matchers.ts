import {
  MatcherHintOptions,
  matcherHint,
  printDiffOrStringify,
  printExpected,
  printReceived,
  stringify,
} from 'jest-matcher-utils';
import * as prettier from 'prettier';

import { poorMansElixirFormat } from './elixir';

const EXPECTED_LABEL = 'Expected';
const RECEIVED_LABEL = 'Received';

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace, @typescript-eslint/no-unused-vars
  namespace jest {
    interface Matchers<R> {
      toHaveFormattedElixirText(test: string): R;
      toHaveFormattedTypeScriptText(text: string): R;
    }
  }
}

type SourceFile = {
  getText(): string;
};

const compareText = (
  received: string,
  expected: string,
  matcherName: string,
): jest.CustomMatcherResult => {
  const matcherOptions: MatcherHintOptions = {
    comment: 'deep equality',
  };
  const pass = received === expected;

  const message = pass
    ? () =>
        // eslint-disable-next-line prefer-template
        matcherHint(matcherName, undefined, undefined, matcherOptions) +
        '\n\n' +
        `Expected: not ${printExpected(expected)}\n` +
        (stringify(expected) !== stringify(received)
          ? `Received:     ${printReceived(received)}`
          : '')
    : () =>
        // eslint-disable-next-line prefer-template
        matcherHint(matcherName, undefined, undefined, matcherOptions) +
        '\n\n' +
        printDiffOrStringify(expected, received, EXPECTED_LABEL, RECEIVED_LABEL, true);

  return { message, pass };
};

export const createMatchers = () => ({
  toHaveFormattedElixirText: (
    sourceFile: SourceFile,
    expectedText: string,
  ): jest.CustomMatcherResult => {
    return compareText(
      poorMansElixirFormat(sourceFile.getText()),
      poorMansElixirFormat(expectedText),
      'toHaveFormattedElixirText',
    );
  },

  toHaveFormattedTypeScriptText: (
    sourceFile: SourceFile,
    expectedText: string,
  ): jest.CustomMatcherResult => {
    const options: prettier.Options = {
      parser: 'typescript',
      singleQuote: true,
      trailingComma: 'all',
    };
    const received = prettier.format(sourceFile.getText(), options);
    const expected = prettier.format(expectedText, options);

    return compareText(received, expected, 'toHaveFormattedTypeScriptText');
  },
});
