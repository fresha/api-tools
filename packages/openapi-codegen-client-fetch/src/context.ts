import type { Nullable } from '@fresha/api-tools-core';
import type { TSProjectContext } from '@fresha/openapi-codegen-utils';
import type { OperationModel } from '@fresha/openapi-model/build/3.0.3';
import type { SourceFile } from 'ts-morph';

export type NamingConvention = 'camel' | 'kebab' | 'snake' | 'title';

export interface Context extends TSProjectContext {
  readonly includeDeprecated: boolean;
  readonly includeInternal: boolean;
  readonly includedTags: Set<string>;
  readonly excludedTags: Set<string>;
  readonly apiNaming: Nullable<NamingConvention>;
  readonly clientNaming: Nullable<NamingConvention>;
}

export interface ActionContext extends Context {
  readonly operation: OperationModel;
  readonly sourceFile: SourceFile;
  readonly typesFile: SourceFile;
}
