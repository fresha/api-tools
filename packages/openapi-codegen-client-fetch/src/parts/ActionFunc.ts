import { camelCase, Nullable, titleCase } from '@fresha/api-tools-core';
import { addFunction, addImportDeclaration, addImportDeclarations } from '@fresha/code-morph-ts';
import {
  assert,
  getOperationIdOrThrow,
  getOperationRequestBodySchema,
  getOperationParameters,
  getOperationResponseSchemas,
  getCodegenOptions,
} from '@fresha/openapi-codegen-utils';
import { CodeBlockWriter, SyntaxKind } from 'ts-morph';

import { DocumentType } from './DocumentType';
import { schemaToType } from './utils';

import type { NamedType } from './NamedType';
import type { ActionContext } from '../context';
import type { ParameterModel } from '@fresha/openapi-model/build/3.0.3';
import type { FunctionDeclaration } from 'ts-morph';

type ParameterInfo = {
  parameter: ParameterModel;
  typeString: string;
};

type VariableName = string; // used to indicate strings that are variable names
type HeaderName = string;

export class ActionFunc {
  readonly context: ActionContext;
  readonly name: string;
  protected readonly parameterVars: Map<VariableName, ParameterInfo>;
  protected readonly pathParametesVars: Map<string, VariableName>;
  protected readonly headerToSet: Map<HeaderName, VariableName>;
  protected requestType: Nullable<NamedType>;
  protected responseType: Nullable<NamedType>;
  protected parsesResponse: boolean;
  // FIXME handle other authorization schemes
  protected usesAuthCookie: boolean;

  constructor(context: ActionContext) {
    this.context = context;
    this.name = getOperationIdOrThrow(this.context.operation);
    this.parameterVars = new Map<string, ParameterInfo>();
    this.pathParametesVars = new Map<string, string>();
    this.headerToSet = new Map<HeaderName, VariableName>();
    this.requestType = null;
    this.responseType = null;
    this.parsesResponse = true;
    this.usesAuthCookie = false;
  }

  collectData(namedTypes: Map<string, NamedType>): void {
    this.context.logger.info(`Collecting data for action ${this.name}`);
    this.collectParameters();
    this.collectRequestData(namedTypes);
    this.collectResponseData(namedTypes);
    this.collectAuthData();
  }

  protected collectParameters(): void {
    for (const parameter of getOperationParameters(this.context.operation)) {
      const paramName = camelCase(parameter.name);
      const typeString = schemaToType(parameter.schema);

      this.parameterVars.set(paramName, {
        typeString,
        parameter,
      });

      if (parameter.in === 'path') {
        this.pathParametesVars.set(parameter.name, paramName);
      }
      if (parameter.in === 'header') {
        this.headerToSet.set(parameter.name, paramName);
      }
    }
  }

  protected collectRequestData(namedTypes: Map<string, NamedType>): void {
    const requestSchema = getOperationRequestBodySchema(
      this.context.operation,
      this.context.useJsonApi,
    );
    if (requestSchema) {
      const requestTypeName = titleCase(`${this.name}Request`);
      assert(
        !namedTypes.has(requestTypeName),
        `Name collision for request type ${requestTypeName}`,
        this.context.operation,
      );
      this.requestType = new DocumentType(
        this.context,
        requestTypeName,
        requestSchema,
        true,
        !!this.context.operation.requestBody?.required,
      );
      this.requestType.collectData(namedTypes);
      namedTypes.set(requestTypeName, this.requestType);
    }
  }

  protected collectResponseData(namedTypes: Map<string, NamedType>): void {
    const responseSchemas = getOperationResponseSchemas(
      this.context.operation,
      this.context.useJsonApi,
    );
    if (responseSchemas.length) {
      const responseTypeName = titleCase(`${this.name}Response`);
      assert(
        !namedTypes.has(responseTypeName),
        `Name collision for request type ${responseTypeName}`,
        this.context.operation,
      );
      // FIXME pass all schemas, not just the first one
      this.responseType = new DocumentType(
        this.context,
        responseTypeName,
        responseSchemas[0],
        false,
        true,
      );
      this.responseType.collectData(namedTypes);
      namedTypes.set(responseTypeName, this.responseType);
    }

    const opts = getCodegenOptions(this.context.operation, 'client-fetch');
    if (opts != null) {
      this.parsesResponse = opts.return !== 'response';
    }
  }

  protected collectAuthData(): void {
    this.usesAuthCookie = !!this.context.operation.getSecurityRequirements().length;
  }

  generateCode(): void {
    this.context.logger.info(`Generating code for action ${this.name}`);

    addImportDeclarations(this.context.sourceFile, {
      './utils': ['COMMON_HEADERS', 'makeUrl', this.parsesResponse ? 'callJsonApi' : 'callApi'],
    });

    if (this.parameterVars.size) {
      addImportDeclaration(this.context.sourceFile, './utils', 'addQueryParam');
    }

    if (this.usesAuthCookie) {
      addImportDeclaration(this.context.sourceFile, './utils', 'authorizeRequest');
    }

    if (this.requestType) {
      addImportDeclaration(this.context.sourceFile, './types', `t:${this.requestType.name}`);
    }
    if (this.responseType) {
      addImportDeclaration(this.context.sourceFile, './types', `t:${this.responseType.name}`);
    }

    let resultType = 'void';
    if (!this.parsesResponse) {
      resultType = 'Response';
    } else if (this.responseType) {
      resultType = this.responseType?.name;
    }

    addImportDeclarations(this.context.sourceFile, {
      './utils': ['t:ExtraCallParams', 'applyExtraParams'],
    });

    const actionFuncHasParams = !!(this.parameterVars.size || this.requestType);
    const actionFunc = addFunction(
      this.context.sourceFile,
      this.name,
      actionFuncHasParams
        ? { params: '{}', extraParams: 'ExtraCallParams' }
        : { extraParams: 'ExtraCallParams' },
      `Promise<${resultType}>`,
      true,
    );
    actionFunc.setIsAsync(true);

    this.generateJsDocs(actionFunc, actionFuncHasParams);
    this.generateParameters(actionFunc);

    actionFunc.addStatements((writer: CodeBlockWriter) => {
      writer.writeLine(`const url = makeUrl(${this.generateCallUrl()});`);

      for (const [paramName, paramInfo] of this.parameterVars) {
        if (paramInfo.parameter.in === 'query') {
          writer.writeLine(
            `addQueryParam(url, '${paramInfo.parameter.name}', params.${paramName});`,
          );
        }
      }
      writer.newLine();

      if (this.requestType) {
        if (this.context.apiNaming) {
          const transformFunc = `${this.context.apiNaming}CaseDeep`;
          addImportDeclaration(this.context.sourceFile, '@fresha/api-tools-core', 't:JSONValue');
          addImportDeclaration(this.context.sourceFile, '@fresha/api-tools-core', transformFunc);
          writer.writeLine(`const body = ${transformFunc}(params.body as unknown as JSONValue)`);
        } else {
          writer.writeLine('const body = params.body;');
        }
      }

      writer.writeLine('const request = {');
      writer.indent(() => {
        const httpMethod = this.context.operation.httpMethod.toUpperCase();
        if (httpMethod !== 'GET') {
          writer.writeLine(`method: '${httpMethod}',`);
        }
        if (this.requestType) {
          writer.writeLine('body: JSON.stringify(body),');
        }
        if (this.headerToSet.size) {
          writer.write('headers:');
          writer.block(() => {
            writer.writeLine('...COMMON_HEADERS,');
            for (const [headerName, varName] of this.headerToSet) {
              writer.writeLine(`'${headerName}': params.${varName}`);
            }
          });
          writer.writeLine(',');
        } else {
          writer.writeLine('headers: COMMON_HEADERS,');
        }
      });
      writer.writeLine('};');
      writer.newLine();

      if (this.usesAuthCookie) {
        writer.writeLine('authorizeRequest(request, extraParams);');
        writer.newLine();
      }

      writer.writeLine('applyExtraParams(request, extraParams)');
      writer.newLine();

      if (this.parsesResponse) {
        if (this.responseType) {
          if (this.context.clientNaming) {
            const transformFunc = `${this.context.clientNaming}CaseDeep`;
            addImportDeclaration(this.context.sourceFile, '@fresha/api-tools-core', transformFunc);
            writer.writeLine(`let response = await callJsonApi(url, request);`);
            writer.newLine();
            writer.writeLine(`response = ${transformFunc}(response);`);
          } else {
            writer.writeLine(`const response = await callJsonApi(url, request);`);
          }
          writer.newLine();
          addImportDeclaration(this.context.sourceFile, './utils', 'dispatchSuccess');
          writer.writeLine(
            `dispatchSuccess('${this.name}', ${
              actionFuncHasParams ? 'params' : 'undefined'
            }, response)`,
          );
          writer.newLine();
          addImportDeclaration(this.context.sourceFile, './utils', 'transformResponse');
          writer.writeLine(
            `return transformResponse<${this.responseType.name}>('${this.name}', response)`,
          );
        } else {
          writer.writeLine(`await callJsonApi(url, request);`);
          writer.newLine();
          writer.writeLine(`return`);
        }
      } else {
        writer.writeLine(`const response = await callApi(url, request);`);
        writer.newLine();
        writer.writeLine(`return response;`);
      }
    });
  }

  protected generateJsDocs(actionFunc: FunctionDeclaration, actionFuncHasParams: boolean): void {
    actionFunc.addJsDoc(writer => {
      let addNL = false;
      if (this.context.operation.summary) {
        writer.write(this.context.operation.summary);
        addNL = true;
      }
      if (this.context.operation.description) {
        if (addNL) {
          writer.newLine();
          writer.newLine();
        }
        writer.write(this.context.operation.description);
        addNL = true;
      }
      if (addNL) {
        writer.newLine();
      }

      if (actionFuncHasParams) {
        writer.newLine();
        writer.writeLine('@param params call parameters');
        if (this.requestType) {
          writer.writeLine('@param params.body request body');
        }
        for (const [paramName, { parameter }] of this.parameterVars) {
          writer.writeLine(
            [
              '@param',
              parameter.required ? `params.${paramName}` : `[params.${paramName}]`,
              parameter.deprecated ? '(deprecated)' : '',
              parameter.description || '',
            ]
              .filter(part => !!part.length)
              .join(' '),
          );
        }
      }

      writer.newLine();
      writer.writeLine('@param [extraParams] additional parameters');
      writer.writeLine(
        '@param [extraParams.authCookieName] name of the authorization cookie for this request',
      );
      writer.writeLine(
        '@param [extraParams.authCookie] value of the authorization cookie for this request',
      );
      writer.writeLine(
        '@param [extraParams.xForwardedFor] sends X-Forwarded-For header with specified value',
      );
      writer.writeLine(
        '@param [extraParams.xForwardedHost] sends X-Forwarded-Host header with specified value',
      );
      writer.write(
        '@param [extraParams.xForwardedProto] sends X-Forwarded-Proto header with specified value',
      );
    });
  }

  protected generateParameters(actionFunc: FunctionDeclaration): void {
    actionFunc.getParameterOrThrow('extraParams').setHasQuestionToken(true);

    if (this.parameterVars.size || this.requestType) {
      const paramsType = actionFunc
        .getParameterOrThrow('params')
        .getTypeNodeOrThrow()
        .asKindOrThrow(SyntaxKind.TypeLiteral);

      if (this.requestType) {
        paramsType.addProperty({
          name: 'body',
          type: this.requestType.name,
          hasQuestionToken: !this.requestType.isRequired,
        });
      }

      for (const [paramName, paramInfo] of this.parameterVars) {
        if (paramInfo.typeString === 'JSONObject') {
          addImportDeclaration(this.context.sourceFile, '@fresha/api-tools-core', 't:JSONObject');
        } else if (paramInfo.typeString === 'JSONArray') {
          addImportDeclaration(this.context.sourceFile, '@fresha/api-tools-core', 't:JSONArray');
        }

        paramsType.addProperty({
          name: paramName,
          type: paramInfo.typeString,
          hasQuestionToken: !paramInfo.parameter.required,
        });
      }
    }
  }

  protected generateCallUrl(): string {
    const { pathUrl } = this.context.operation.parent;
    const str = pathUrl.replace(/\{\w+\}/g, match => {
      const argName = this.pathParametesVars.get(match.slice(1, -1));
      assert(
        argName,
        `Found '${match}' parameter in '${pathUrl}', but not path parameter`,
        this.context.operation,
      );
      return `\${params.${argName}}`;
    });

    return `\`${str}\``;
  }
}
