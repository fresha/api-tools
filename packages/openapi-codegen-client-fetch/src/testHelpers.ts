// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';

import type { ActionContext, Context } from './context';
import type { OpenAPIModel, OperationModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel): Context => {
  const base = createTSProjectTestContext(openapi);

  return {
    ...base,
    includeDeprecated: false,
    includedTags: new Set<string>(),
    excludedTags: new Set<string>(),
    transformRequest: 'kebab',
    transformResponse: 'camel',
  };
};

export const createActionTestContext = (
  operation: OperationModel,
  fileName: string,
): ActionContext => {
  const base = createTestContext(operation.root);
  const sourceFile = base.project.createSourceFile(fileName, '');

  return {
    ...base,
    sourceFile,
    operation,
  };
};
