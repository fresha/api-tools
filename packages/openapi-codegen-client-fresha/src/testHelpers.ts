// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { getAPIName, getRootUrlOrThrow } from '@fresha/openapi-codegen-utils';

import type { Context } from './context';
import type { OpenAPIModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel, rootDir = '/'): Context => {
  const base = createTSProjectTestContext(openapi, rootDir);

  let apiRootUrl: string;
  try {
    apiRootUrl = getRootUrlOrThrow(openapi);
  } catch (err) {
    apiRootUrl = 'API_URL';
  }

  return {
    ...base,
    apiName: getAPIName(openapi),
    apiRootUrl,
  };
};
