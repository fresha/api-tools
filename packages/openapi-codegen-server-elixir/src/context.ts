import type { Project } from '@fresha/code-morph-ex';
import type { JSONAPISchemaRegistry } from '@fresha/json-api-model';
import type { Context as BasicContext } from '@fresha/openapi-codegen-utils';

export interface Context extends BasicContext {
  readonly testObjectFactoryModuleName: string;
  readonly project: Project;
  readonly registry: JSONAPISchemaRegistry;
}
