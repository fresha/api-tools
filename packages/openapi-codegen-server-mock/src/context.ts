export { createTSProjectContext as createContext } from '@fresha/openapi-codegen-utils';

export interface CreateContextParams {
  input: string;
  output: string;
  jsonApi?: boolean;
  verbose?: boolean;
  dryRun?: boolean;
}
