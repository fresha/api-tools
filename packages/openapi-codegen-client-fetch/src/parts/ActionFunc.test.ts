import { buildEmployeeSchemasForTesting } from '@fresha/openapi-codegen-test-utils';
import {
  addResourceAttributes,
  addResourceRelationships,
  MEDIA_TYPE_JSON_API,
  RelationshipCardinality,
  setResourceSchema,
} from '@fresha/openapi-codegen-utils';
import { OpenAPIFactory, OperationModel, SchemaFactory } from '@fresha/openapi-model/build/3.0.3';

import '@fresha/openapi-codegen-test-utils/build/matchers';

import { ActionContext } from '../context';
import { createActionTestContext } from '../testHelpers';

import { ActionFunc } from './ActionFunc';

import type { NamedType } from './NamedType';

const createAction = (operation: OperationModel, filePath = 'src/index.ts'): ActionFunc => {
  const context = createActionTestContext(operation, filePath);
  return new ActionFunc(context);
};

test('simple test', () => {
  const openapi = OpenAPIFactory.create();
  buildEmployeeSchemasForTesting(openapi);

  openapi.components.setSecuritySchema('the_auth', 'apiKey');

  const operation = openapi.setPathItem('/employees').setOperation('get');
  operation.operationId = 'readEmployeeList';

  const paramOffset = operation.addParameter('offset', 'query');
  paramOffset.required = true;
  paramOffset.schema = SchemaFactory.create(paramOffset, 'number');

  const paramLimit = operation.addParameter('limit', 'query');
  paramLimit.required = false;
  paramLimit.schema = SchemaFactory.create(paramLimit, 'number');

  operation
    .setResponse(200, 'returns a list of employees')
    .setContent(MEDIA_TYPE_JSON_API)
    .setSchema('object')
    .setProperty('data', 'array')
    .setItems(openapi.components.getSchemaOrThrow('EmployeeResource'));
  operation.addSecurityRequirement().addScopes('the_auth');

  const namedTypes = new Map<string, NamedType>();
  const generatedTypes = new Set<string>();

  const action = createAction(operation);
  action.collectData(namedTypes);
  action.generateCode(generatedTypes);

  expect(action.context.project.getSourceFile('src/index.ts')).toHaveFormattedTypeScriptText(`
    import {
      COMMON_HEADERS,
      makeUrl,
      callJsonApi,
      addQueryParam,
      authorizeRequest,
      ExtraCallParams,
      applyExtraParams,
      dispatchSuccess,
    } from './utils';
    import type {
      JSONAPIServerResource,
      JSONAPIResourceRelationship1,
      JSONAPIResourceRelationship0,
      JSONAPIResourceRelationshipN,
      JSONAPIDataDocument,
    } from "@fresha/api-tools-core";

    export type EmployeeResource = JSONAPIServerResource<
      'employees',
      {
        name: string;
        age?: number | null;
        gender?: 'male' | 'female' | 'other' | null;
        active?: boolean;
      },
      {
        manager: JSONAPIResourceRelationship1<'employees'>;
        buddy: JSONAPIResourceRelationship0<'employees'>;
        subordinates: JSONAPIResourceRelationshipN<'employees'>;
      }
    >;

    export type ReadEmployeeListResponse = JSONAPIDataDocument<EmployeeResource[]>;

    export async function readEmployeeList(
      params: {
        offset: number,
        limit?: number,
      },
      extraParams?: ExtraCallParams,
    ): Promise<ReadEmployeeListResponse> {
      const url = makeUrl(\`/employees\`);
      addQueryParam(url, 'offset', params.offset);
      addQueryParam(url, 'limit', params.limit);

      const request = {
        headers: COMMON_HEADERS,
      };

      authorizeRequest(request, extraParams);

      applyExtraParams(request, extraParams);

      const response = await callJsonApi(url, request);

      dispatchSuccess('readEmployeeList', params, response);

      return response as unknown as ReadEmployeeListResponse;
    }
  `);
});

test('specific naming convention for client library', () => {
  const openapi = OpenAPIFactory.create();

  // construct a resource with kebab-case attributes
  const employee = openapi.components.setSchema('EmployeeResource');
  setResourceSchema(employee, 'employees');
  addResourceAttributes(employee, {
    'old-name': { type: 'string', required: true },
    'new-name': { type: 'string', required: true },
    age: { type: 'number', nullable: true },
    'is-active': { type: 'boolean', nullable: false },
  });
  addResourceRelationships(employee, {
    'direct-manager': { resourceType: 'employees', cardinality: RelationshipCardinality.One },
  });

  openapi.components.setSecuritySchema('the_auth', 'apiKey');

  const operation = openapi.setPathItem('/employees').setOperation('patch');
  operation.operationId = 'readEmployeeList';

  operation
    .setResponse(200, 'returns a list of employees')
    .setContent(MEDIA_TYPE_JSON_API)
    .setSchema('object')
    .setProperty('data', 'array')
    .setItems(openapi.components.getSchemaOrThrow('EmployeeResource'));
  operation.addSecurityRequirement().addScopes('the_auth');

  const namedTypes = new Map<string, NamedType>();
  const generatedTypes = new Set<string>();

  // override default values for api/client naming conventions
  const context: ActionContext = {
    ...createActionTestContext(operation, 'src/index.ts'),
    apiNaming: 'kebab',
    clientNaming: 'camel',
  };
  const action = new ActionFunc(context);

  action.collectData(namedTypes);
  action.generateCode(generatedTypes);

  expect(action.context.project.getSourceFile('src/index.ts')).toHaveFormattedTypeScriptText(`
    import {
      COMMON_HEADERS,
      makeUrl,
      callJsonApi,
      authorizeRequest,
      ExtraCallParams,
      applyExtraParams,
      dispatchSuccess,
    } from './utils';
    import {
      JSONAPIServerResource,
      JSONAPIResourceRelationship1,
      JSONAPIDataDocument,
      camelCaseDeep,
    } from "@fresha/api-tools-core";

    export type EmployeeResource = JSONAPIServerResource<
      'employees',
      {
        oldName: string;
        newName: string;
        age?: number | null;
        isActive?: boolean;
      },
      {
        directManager: JSONAPIResourceRelationship1<'employees'>;
      }
    >;

    export type ReadEmployeeListResponse = JSONAPIDataDocument<EmployeeResource[]>;

    export async function readEmployeeList(
      extraParams?: ExtraCallParams,
    ): Promise<ReadEmployeeListResponse> {
      const url = makeUrl(\`/employees\`);

      const request = {
        method: 'PATCH',
        headers: COMMON_HEADERS,
      };

      authorizeRequest(request, extraParams);

      applyExtraParams(request, extraParams);

      let response = await callJsonApi(url, request);

      response = camelCaseDeep(response);

      dispatchSuccess('readEmployeeList', undefined, response);

      return response as unknown as ReadEmployeeListResponse;
    }
  `);
});

test('action returns raw response', () => {
  const openapi = OpenAPIFactory.create();

  const schema = openapi.components.setSchema('EmployeeResource');
  setResourceSchema(schema, 'employees');

  const operation = openapi.setPathItem('/employees').setOperation('get');
  operation.operationId = 'readEmployeeList';
  operation
    .setResponse(200, 'returns a list of employees')
    .setContent(MEDIA_TYPE_JSON_API)
    .setSchema('object')
    .setProperty('data', 'array')
    .setItems(schema);
  operation.setExtension('fresha-codegen', {
    'client-fetch': {
      return: 'response',
    },
  });

  const namedTypes = new Map<string, NamedType>();
  const generatedTypes = new Set<string>();

  const action = createAction(operation);
  action.collectData(namedTypes);
  action.generateCode(generatedTypes);

  expect(action.context.project.getSourceFile('src/index.ts')).toHaveFormattedTypeScriptText(`
    import {
      COMMON_HEADERS,
      makeUrl,
      callApi,
      ExtraCallParams,
      applyExtraParams
    } from './utils';
    import type {
      JSONAPIServerResource,
      JSONAPIDataDocument,
    } from "@fresha/api-tools-core";

    export type EmployeeResource = JSONAPIServerResource<'employees'>;

    export type ReadEmployeeListResponse = JSONAPIDataDocument<EmployeeResource[]>;

    export async function readEmployeeList(
      extraParams?: ExtraCallParams,
    ): Promise<Response> {
      const url = makeUrl(\`/employees\`);

      const request = {
        headers: COMMON_HEADERS,
      };

      applyExtraParams(request, extraParams);

      const response = await callApi(url, request);

      return response;
    }
  `);
});
