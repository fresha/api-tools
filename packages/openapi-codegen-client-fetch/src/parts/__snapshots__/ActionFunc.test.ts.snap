// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action returns raw response: src/index.ts 1`] = `
"import { COMMON_HEADERS, makeUrl, callApi, ExtraCallParams, applyExtraParams } from "./utils";
import type { ReadEmployeeListResponse } from "./types";

/**
 * @param [extraParams] additional parameters
 * @param [extraParams.authCookieName] name of the authorization cookie for this request
 * @param [extraParams.authCookie] value of the authorization cookie for this request
 * @param [extraParams.xForwardedFor] sends X-Forwarded-For header with specified value
 * @param [extraParams.xForwardedHost] sends X-Forwarded-Host header with specified value
 * @param [extraParams.xForwardedProto] sends X-Forwarded-Proto header with specified value
 */
export async function readEmployeeList(extraParams?: ExtraCallParams): Promise<Response> {
    const url = makeUrl(\`/employees\`);

    const headerParameters: Record<string, string> = {};

    const request = {
        headers: {...COMMON_HEADERS, ...headerParameters},
    };

    applyExtraParams(request, extraParams)

    const response = await callApi(url, request);

    return response;
}
"
`;

exports[`action returns raw response: src/types.ts 1`] = `
"import type { JSONAPIServerResource, JSONAPIDataDocument } from "@fresha/api-tools-core";


/**
 * EmployeeResource
 *
 */
export type EmployeeResource = JSONAPIServerResource<'employees'>;

/**
 * This is a response resource
 *
 * Primary data resources:
 *  - EmployeeResource
 *
 */
export type ReadEmployeeListResponse = JSONAPIDataDocument<(EmployeeResource)[]>;
"
`;

exports[`simple test: src/index.ts 1`] = `
"import { COMMON_HEADERS, makeUrl, callJsonApi, addQueryParam, authorizeRequest, ExtraCallParams, applyExtraParams, dispatchSuccess, transformResponse } from "./utils";
import type { ReadEmployeeListResponse } from "./types";

/**
 * Reads employee list
 *
 * This operation reads a list of employee resources, and returns them as a JSON:API
 * document. Besides employees, it includes resources associated with employees, like
 * addresses and **departments**.
 *
 * @param params call parameters
 * @param params.offset Index of the first employee to retrieve
 * @param [params.limit] Maximal number of employee resources to retrieve
 *
 * @param [extraParams] additional parameters
 * @param [extraParams.authCookieName] name of the authorization cookie for this request
 * @param [extraParams.authCookie] value of the authorization cookie for this request
 * @param [extraParams.xForwardedFor] sends X-Forwarded-For header with specified value
 * @param [extraParams.xForwardedHost] sends X-Forwarded-Host header with specified value
 * @param [extraParams.xForwardedProto] sends X-Forwarded-Proto header with specified value
 */
export async function readEmployeeList(params: {
    offset: number;
    limit?: number;
}, extraParams?: ExtraCallParams): Promise<ReadEmployeeListResponse> {
    const url = makeUrl(\`/employees\`);
    addQueryParam(url, 'offset', params.offset);
    addQueryParam(url, 'limit', params.limit);

    const headerParameters: Record<string, string> = {};

    const request = {
        headers: {...COMMON_HEADERS, ...headerParameters},
    };

    authorizeRequest(request, extraParams);

    applyExtraParams(request, extraParams)

    const response = await callJsonApi(url, request);

    dispatchSuccess('readEmployeeList', params, response)

    return transformResponse<ReadEmployeeListResponse>('readEmployeeList', response)
}
"
`;

exports[`simple test: src/types.ts 1`] = `
"import type { JSONAPIServerResource, JSONAPIResourceRelationship1, JSONAPIResourceRelationship0, JSONAPIResourceRelationshipN, JSONAPIDataDocument } from "@fresha/api-tools-core";


/**
 * EmployeeResource
 *
 * Attributes:
 * -  name
 * -  age
 * -  gender
 * -  active
 *
 * Relationships:
 * -  manager relationship to the 'employees' resource
 * -  buddy relationship to the 'employees' resource
 * -  subordinates relationship to the 'employees' resource
 *
 */
export type EmployeeResource = JSONAPIServerResource<'employees', {
        name: string;
        age?: number | null;
        gender?: 'male' | 'female' | 'other' | null;
        active?: boolean;
}, {
        manager: JSONAPIResourceRelationship1<'employees'>;
        buddy: JSONAPIResourceRelationship0<'employees'>;
        subordinates: JSONAPIResourceRelationshipN<'employees'>;
}>;

/**
 * This is a response resource
 *
 * Primary data resources:
 *  - EmployeeResource
 *
 */
export type ReadEmployeeListResponse = JSONAPIDataDocument<(EmployeeResource)[]>;
"
`;

exports[`specific naming convention for client library: src/index.ts 1`] = `
"import { COMMON_HEADERS, makeUrl, callJsonApi, authorizeRequest, ExtraCallParams, applyExtraParams, dispatchSuccess, transformResponse } from "./utils";
import type { ReadEmployeeListResponse } from "./types";
import { camelCaseDeep } from "@fresha/api-tools-core";

/**
 * @param [extraParams] additional parameters
 * @param [extraParams.authCookieName] name of the authorization cookie for this request
 * @param [extraParams.authCookie] value of the authorization cookie for this request
 * @param [extraParams.xForwardedFor] sends X-Forwarded-For header with specified value
 * @param [extraParams.xForwardedHost] sends X-Forwarded-Host header with specified value
 * @param [extraParams.xForwardedProto] sends X-Forwarded-Proto header with specified value
 */
export async function readEmployeeList(extraParams?: ExtraCallParams): Promise<ReadEmployeeListResponse> {
    const url = makeUrl(\`/employees\`);

    const headerParameters: Record<string, string> = {};

    const request = {
        method: 'PATCH',
        headers: {...COMMON_HEADERS, ...headerParameters},
    };

    authorizeRequest(request, extraParams);

    applyExtraParams(request, extraParams)

    let response = await callJsonApi(url, request);

    response = camelCaseDeep(response);

    dispatchSuccess('readEmployeeList', undefined, response)

    return transformResponse<ReadEmployeeListResponse>('readEmployeeList', response)
}
"
`;

exports[`specific naming convention for client library: src/types.ts 1`] = `
"import type { JSONAPIServerResource, JSONAPIResourceRelationship1, JSONAPIDataDocument } from "@fresha/api-tools-core";


/**
 * EmployeeResource
 *
 * Attributes:
 * -  old-name
 * -  new-name
 * -  age
 * -  is-active
 *
 * Relationships:
 * -  direct-manager relationship to the 'employees' resource
 *
 */
export type EmployeeResource = JSONAPIServerResource<'employees', {
        oldName: string;
        newName: string;
        age?: number | null;
        isActive?: boolean;
}, {
        directManager: JSONAPIResourceRelationship1<'employees'>;
}>;

/**
 * This is a response resource
 *
 * Primary data resources:
 *  - EmployeeResource
 *
 */
export type ReadEmployeeListResponse = JSONAPIDataDocument<(EmployeeResource)[]>;
"
`;

exports[`test optional header parameters: src/index.ts 1`] = `
"import { COMMON_HEADERS, makeUrl, callJsonApi, authorizeRequest, ExtraCallParams, applyExtraParams } from "./utils";

/**
 * operation with an optional header
 *
 * This operation has an optional header parameter
 *
 * @param params call parameters
 * @param [params.OptionalHeaderParameter] optional header parameter
 *
 * @param [extraParams] additional parameters
 * @param [extraParams.authCookieName] name of the authorization cookie for this request
 * @param [extraParams.authCookie] value of the authorization cookie for this request
 * @param [extraParams.xForwardedFor] sends X-Forwarded-For header with specified value
 * @param [extraParams.xForwardedHost] sends X-Forwarded-Host header with specified value
 * @param [extraParams.xForwardedProto] sends X-Forwarded-Proto header with specified value
 */
export async function test(params: {
    OptionalHeaderParameter?: string;
}, extraParams?: ExtraCallParams): Promise<void> {
    const url = makeUrl(\`/test-endpoint\`);

    const headerParameters: Record<string, string> = {};

    if (params.OptionalHeaderParameter) headerParameters['Optional-Header-Parameter'] = params.OptionalHeaderParameter

    const request = {
        headers: {...COMMON_HEADERS, ...headerParameters},
    };

    authorizeRequest(request, extraParams);

    applyExtraParams(request, extraParams)

    await callJsonApi(url, request);

    return
}
"
`;
